packages/common/container-definitions/src/loader.ts

```diff
@@ -583,6 +587,11 @@ export interface IContainerLoadMode {
 	 * Default value.
 	 */
 	| undefined
+		/*
+		 * Only fetch and apply trailing ops up until (and including) the specified sequence number.
+		 * Requires `ILoaderHeader["fluid-sequence-number"]` to also be defined.
+		 */
+		| "sequenceNumber"
```

Comment:
I'll leave comment here (to allow proper commenting thread), but it applies to whole PR.
I'd love to see an exploration where same functionality is exposed as proper APIs on Container / IContainer.
I do not think complicating matrix of options is a good direction. I'll raise some of the concerns I see:
1. loadToSequenceNumber:
   - what if this number is higher than current sequence number when client is fully up-to-date?
   - what if it's below snapshot sequence number?
   - in both cases I'd expect user of API to have to deal / decide things things. But the form of API results in most customers not thinking about these cases
2. Freezing needs to have unfreezing API counterpart. Otherwise, it's not clear how to use it (outside very specific scenario you want to use it)
3. Why we disconnect after freeze? What if another scenario exists that want the same, but not have disconnection (i.e. eventually unfreeze and keep moving forward)?

General observations:
1. Instead of piling up, I'd rather see solution that removes any references to sequence numbers from options, and provides flat (stand-alone) API that takes container object allows waiting for specific sequence number.
   - I expect to see different implementations here with different expectations / assumptions (i.e. once needs to get to >= sequence number, one wants specific number and will throw if it can't get there, etc.)
2. I'd love to see pause / unpause APIs on container that can be used for this scenario, but also for any other scenarios.
3. Connect/disconnect APIs already exists, no need to add another way to get there.

Reply:
Re: an API that takes a container object - what strategy are you thinking for getting a container reference back before processing beyond that sequence number?  Something like loading with deltaConnection === "none" and explicitly calling connect() after attaching listeners to the deltaManager?

Even if we can make a flow like that work, I also think it's limiting to have customers pulling the deltaManager reference on IContainer - yes they do it today, but if we ever want to move the deltaManager to the runtime it'll likely become inappropriate to access it through the IContainer at that time.

I don't think the Container/DeltaManager's construction and layering permit a perfect solution today.  But once a better alternative becomes available I think we can provide it and move users over to it.

Reply:
I do not know what "move IDeltaManager to container runtime layers" means (in practical terms) and if it's a good idea, so I'm not sure I'd mix it here. If we chose to do so, it would create same kind of problem (trouble controlling state of Container RE how much it is allowed to progress) no matter what type of API we have for controlling it (pause vs. process up to X ops) and would likely require a custom hand-shake between caller and container runtime. So I'd table this subject and focus on current layering.

If you are adding paused state, then I believe you need pause/unpause APIs. I really do not want to see here APIs that only could be used for one thing.

One way or another we will have a bunch of smallish tools here. We already have some of these tools (like do not process ops). I think we need to do a better job of describing all of them (across multiple approaches), how well they would work with each other, and how generic they are (vs. being very specialized or with very custom policies). For example, pausing is rather generic. Process ops up to X ops could be generic, but once we start adding very custom policies to make it work (like throw if you are past X) it becomes not great. 

While I'm not fully convinced in this direction, I'm strongly considering (for other reason) to ditch default mode and move toward - socket connection is explicitly controlled by host and could be done only after container is loaded. That's the mode for all of our 1st party clients use. And host can connect/disconnect, freeze / unfreeze (pause/unpause) container as they see fit. Even if we do not do that, we already have API to load with / without ops, with/without connection, and we need pause/unpause, so there does not seem like a need for "load up to X ops" - this flow could be accomplished using all the tools above, moving policy (of what to do if you can't hit X) outside of Loader.

Reply:
@vladsud I'm not following what specific change you're proposing here - could you please clarify?